### ISSUE TITLE
State files created with `tf state mv...` are invalid?

### ISSUE THREAD

**Issue by eyalzek** [hashicorp/terraform#8263]
Up until now we have had a single state file holding 3 environments. This caused a bunch of problems in the past and we planned on separating them to different files for a while. With the release of TF 0.7, the `terraform state` command has been introduced so I thought I'll give it shot. Our old state file was `terraform.tfstate`, I issued the following commands:

`tf state mv -state-out=staging.tfstate module.staging module.staging`
`tf state mv -state-out=production.tfstate module.production module.production`
`tf state mv -state-out=dev.tfstate module.dev module.dev`

I was surprised by how simple it was, the state files themselves looked proper (to my untrained eye). The problem is that now terraform is trying to recreate all of our resources, as if it's not taking the state file into consideration, eg:

`tf plan -state=staging.tfstate -target=module.staging.module.foo`

will plan to create all of the resources defined in the submodule `foo`. The same happens for all the environments, when trying to plan and apply a submodule. I tested again with the old statefile just to make sure and I got the expected output (nothing to be changed).
I attempted refreshing the state, but that didn't change anything, more than that, it added state from other environments to the state file I was trying to refresh.

### Terraform Version
0.7

If any more information is required, I'll be glad to supply it.

**Comment by eyalzek** [comment]
On a second look, it seems that something is missing from the newly created state files. In an older one, I have an entry for every submodule, e.g:

```
"path": [
                "root",
                "staging",
                "foo"
  ],
"outputs": {},
"resources": {*foo submodule resources*}
```

while on the new separate state file I have only a single entry:

```
"path": [
                "root",
                "staging"
  ],
"outputs": {},
"resources": {*the entire env module resources*}
```

which leads me to believe I might have done it wrong. Do I need to go through one submodule at a time and move it? What would be the flow exactly?

**Comment by eyalzek** [comment]
I've wrote a simple bash script to try and move the state to a separate file, looping on all relevant resource:

```
#!/bin/bash -ex
ENV=production
for resource in `terraform state list -state=terraform.tfstate |grep $ENV`
do
    echo $resource
    terraform state mv -state=terraform.tfstate -state-out="$ENV-test.tfstate" $resource $resource
done
```

The structure is not entirely the same though, so I'm still a bit hesitant to commit it. Any input?

**Comment by mitchellh** [comment]
I've reproduced this in a test case and should have a fix out today.

**Comment by eyalzek** [comment]
@mitchellh thanks! In the meantime, would the bash script I put up there give me the same output as expected? Because from what I tested it worked ok.

**Comment by mitchellh** [comment]
Oops sorry I was on an unrefreshed page without internet so I don't think the GitHub auto-loading-new-comments worked and I missed that comment. Yes, that would work and is effectively the bug that is happening. :)

**Comment by mitchellh** [comment]
A fix is queued up in PR #8304! Thanks for the detailed bug report, it helped track this down super fast.
